arraycontext/impl/pyopencl/taggable_cl_array.py

```diff
@@ -0,0 +1,112 @@
+"""
+.. autoclass:: TaggableCLArray
+.. autoclass:: Axis
+
+.. autofunction:: to_tagged_cl_array
+"""
+
+import pyopencl.array as cla
+from typing import FrozenSet, Union, Sequence, Optional, Tuple
+from pytools.tag import Taggable, Tag, tag_dataclass
+
+
+@tag_dataclass
+class Axis(Taggable):
+    """
+    Records the tags corresponding to a dimensions of :class:`TaggableCLArray`.
+    """
+    tags: FrozenSet[Tag]
+
+    def copy(self, **kwargs):
+        from dataclasses import replace
+        return replace(self, **kwargs)
+
+
+class TaggableCLArray(cla.Array, Taggable):
+    """
+    A :class:`pyopencl.array.Array` with additional attributes record
+    """
+    def __init__(self, cq, shape, dtype, order="C", allocator=None,
+                 data=None, offset=0, strides=None, events=None, _flags=None,
+                 _fast=False, _size=None, _context=None, _queue=None,
+                 axes=None, tags=frozenset()):
+
+        super().__init__(cq=cq, shape=shape, dtype=dtype,
+                         order=order, allocator=allocator,
+                         data=data, offset=offset,
+                         strides=strides, events=events,
+                         _flags=_flags, _fast=_fast,
+                         _size=_size, _context=_context,
+                         _queue=_queue)
+
+        self.tags = tags
+        axes = axes if axes is not None else tuple(Axis(frozenset())
+                                                   for _ in self.shape)
```

Comment:
I think it might be better to expect this to be in place, cf. `to_tagged_cl_array`, in  order to keep this constructor fast. It was already one of the hottest code paths in eager eval.

Reply:
Sorry, I don't think that's possible. After doing any pyopencl operation we instantiate a new TaggableCLArray which would mean that we would carry along a `None` for the axes. Tried to keep the costs low by using `pytools.memoize`.

Reply:
OK, I'll take this. At a later point, we may decide that we can save some function calls by inlining, but let's wait with that.